pkg/pgmodel/new_migrate.go

```diff
@@ -0,0 +1,161 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package pgmodel
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/blang/semver/v4"
+	"github.com/jackc/pgx/v4"
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/migrations"
+	"github.com/timescale/promscale/pkg/pgmodel/common/errors"
+	"github.com/timescale/promscale/pkg/pgmodel/common/extension"
+	"github.com/timescale/promscale/pkg/util"
+	"github.com/timescale/promscale/pkg/version"
+)
+
+var (
+	MigrationLockError = fmt.Errorf("Could not acquire migration lock. Ensure there are no other connectors running and try again.")
+	migrateMutex       = &sync.Mutex{}
+)
+
+func doesSchemaMigrationTableExist(db *pgx.Conn) (exists bool, err error) {
+	const stmt = "SELECT count(*) FILTER (WHERE tablename = 'prom_schema_migrations') > 0 FROM pg_tables"
+	err = db.QueryRow(
+		context.Background(),
+		stmt,
+	).Scan(&exists)
+	return
+}
+
+func removeOldExtensionIfExists(db *pgx.Conn) error {
+	const (
+		// transition is the first version of the extension that does the
+		// migrations the new way (i.e. in the extension rather than from promscale connector)
+		transition = "0.5.0"
+		stmt       = "DROP EXTENSION IF EXISTS promscale" // TODO to cascade or not to cascade?
+	)
+
+	installedVersion, installed, err := extension.FetchInstalledExtensionVersion(db, "promscale")
+	if err != nil {
+		return err
+	}
+
+	if installed && installedVersion.LT(semver.MustParse(transition)) {
+		_, err := db.Exec(
+			context.Background(),
+			stmt,
+		)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func installExtensionAllBalls(db *pgx.Conn) error {
+	const stmt = "CREATE EXTENSION promscale SCHEMA _prom_ext VERSION '0.0.0'"
+	_, err := db.Exec(
+		context.Background(),
+		stmt,
+	)
+	return err
+}
+
+func Migrate(conn *pgx.Conn, appVersion VersionInfo, leaseLock *util.PgAdvisoryLock, extOptions extension.ExtensionMigrateOptions) error {
+	// At startup migrators attempt to grab the schema-version lock. If this
+	// fails that means some other connector is running. All is not lost: some
+	// other connector may have migrated the DB to the correct version. We warn,
+	// then start the connector as normal. If we are on the wrong version, the
+	// normal version-check code will prevent us from running.
+
+	if leaseLock != nil {
+		locked, err := leaseLock.GetAdvisoryLock()
+		if err != nil {
+			return fmt.Errorf("error while acquiring migration lock %w", err)
+		}
+		if !locked {
+			return MigrationLockError
+		}
+		defer func() {
+			_, err := leaseLock.Unlock()
+			if err != nil {
+				log.Error("msg", "error while releasing migration lock", "err", err)
+			}
+		}()
+	} else {
+		log.Warn("msg", "skipping migration lock")
+	}
+
+	migrateMutex.Lock()
+	defer migrateMutex.Unlock()
+
```

Comment:
let's check the right extension  version is available here before going through the rest of the migration logic. I'd like to fail before we make any db changes

Reply:
Ahh. Good idea.

Reply:
This commit should address it. https://github.com/timescale/promscale/pull/1132/commits/e3a608353bfe11fc3cce63617c00d51946d47fdb